src/ioctl/mod.rs

```diff
@@ -0,0 +1,295 @@
+//! Unsafe `ioctl` API.
+//!
+//! Unix systems expose a number of `ioctl`'s. Although they were originally meant to modify the
+//! behavior of files, they've now been adopted as a general purpose system call for making calls
+//! into the kernel. In addition to the wide variety of system calls that are included by default
+//! in the kernel, many drivers expose their own `ioctl`'s for controlling their behavior, some
+//! of which are proprietary.  Therefore it is impossible to make a safe interface for every `ioctl`
+//! call, as they all have wildly varying semantics.
+//!
+//! This module provides an unsafe interface to write your own `ioctl` API. To start, create a type
+//! that implements [`Ioctl`]. Then, pass it to [`ioctl`] to make the `ioctl` call.
+
+#![allow(unsafe_code)]
+
+use crate::backend::c;
+use crate::fd::{AsFd, BorrowedFd};
+use crate::io::Result;
+
+#[cfg(any(linux_kernel, apple, bsd))]
+use core::mem;
+
+pub use patterns::*;
+
+mod patterns;
+
+#[cfg(linux_kernel)]
+mod linux;
+
+#[cfg(any(apple, bsd))]
+mod bsd;
+
+#[cfg(linux_kernel)]
+use linux as platform;
+
+#[cfg(any(apple, bsd))]
+use bsd as platform;
+
+/// Perform an `ioctl` call.
+///
+/// `ioctl` was originally intended to act as a way of modifying the behavior of files, but has
+/// since been adopted as a general purpose system call for making calls into the kernel. In
```

Comment:
My sense is that this context about `ioctl` originally being intended for modifying files isn't needed here. These days, `ioctl` just is a general-purpose system call.